scripts/create_deprecated_modules.lean

```diff
@@ -0,0 +1,378 @@
+-/
+
+open Lean Elab Command
+
+namespace DeprecatedModule
+
+/--
+This file interacts with `git ...` quite a bit. `runCmd` takes as input the command-line
+function `git ...` and returns its stdout string as its output, although technically the
+command need not be `git ...`, this is the only situation where we use it.
+
+This is convenient to get both the output of the function, but also for reproducing the exact
+command-line text that produced the output for better reproducibility and error reporting.
+-/
+def runCmd (s : String) : IO String := do
+  let cmd::args := s.splitOn | EStateM.throw "Please provide at least one word in your command!"
+  IO.Process.run {cmd := cmd, args := args.toArray}
+
+/--
+`getHeader fname fileContent keepTrailing` takes as input two strings and a `Bool`ean.
+It uses
+* `fname`, as the path of a file (which need not exist);
+* `fileContent`, as the content of `fname` (regardless of whether the file exists and what its
+  content is);
+* `keepTrailing` a boolean to control whether to keep trailing comments.
+
+It returns just the imports of `fileContent`, including trailing comments if `keepTrailing = true`
+(the command always trims trailing whitespace after the last comment).
+-/
+def getHeader (fname fileContent : String) (keepTrailing : Bool) : IO String := do
+  let (stx, _) ← Parser.parseHeader (Parser.mkInputContext fileContent fname)
+  let stx := if keepTrailing then stx.raw else stx.raw.unsetTrailing
+  let some substring := stx.getSubstring? | throw <| .userError "No substring: we have a problem!"
+  return substring.toString
+
+/--
+`getHeaderFromFileName fname keepTrailing` is similar to `getHeader`, except that it assumes that
+`fname` is the actual path of a file, and uses `fname`'s content as input to `getHeader`.
+-/
+def getHeaderFromFileName (fname : String) (keepTrailing : Bool) : IO String := do
+  getHeader fname (← IO.FS.readFile fname) keepTrailing
+
+/--
+`mkDeprecationWithDate date customMessage` returns the formatted syntax
+`deprecated_module "customMessage" (since := "date")`,
+where the date is of the form `YYYY-MM-DD`.
+-/
+def mkDeprecationWithDate (date : String)
+    (customMessage : Option String := some "Auto-generated deprecation") :
+    CommandElabM Format := do
+  let msgStx := customMessage.map Syntax.mkStrLit
+  let dateStx := Syntax.mkStrLit date
+  let stx ← `(command|deprecated_module $[$msgStx]? (since := $dateStx))
+  liftCoreM <| PrettyPrinter.ppCategory `command stx
+
+/--
+`mkDeprecation customMessage` returns the formatted syntax
+`deprecated_module "customMessage" (since := "YYYY-MM-DD")`, where the date is today's date.
+-/
+def mkDeprecation (customMessage : Option String := some "Auto-generated deprecation") :
+    CommandElabM Format := do
+  -- Get the current date in UTC: we don't want this to depend on the user computer's time zone.
+  let date := s!"{(← Std.Time.DateTime.now (tz := .UTC)).toPlainDate}"
```

Comment:
More precisely, this takes a *current date*: I'd argue this is the wrong thing in applications; we want to take the *commit date* and not the current date!

Reply:
I was aiming for non-retroactive deprecations.  Or, another way of saying, I was giving the benefit of having 6 months where a deprecation warning is active, so, if the deprecation was applied late, it would still last for at least 6 months.